datahub-web-react/src/app/ingest/source/builder/RecipeForm/constants.ts

```diff
@@ -0,0 +1,312 @@
+import { SNOWFLAKE } from '../../conf/snowflake/snowflake';
+
+export enum FieldType {
+    TEXT,
+    BOOLEAN,
+    LIST,
+}
+
+export interface RecipeField {
+    name: string;
+    label: string;
+    tooltip: string;
+    type: FieldType;
+    rules: any[] | null;
+    section?: string;
+    getValueFromRecipe: (recipe: any) => any;
+    setValueOnRecipe: (recipe: any, value: any) => any;
+}
+
+function updateParentAndField(recipe: any, field: string, parentField: string) {
+    const updatedRecipe = { ...recipe };
+    // remove this field from the recipe if it's empty
+    if (updatedRecipe.source.config?.[parentField]?.[field]) {
+        updatedRecipe.source.config[parentField][field] = undefined;
+
+        // remove its parent if there are no other fields underneath it
+        const parentKeys = Object.keys(updatedRecipe.source.config[parentField]);
+        if (parentKeys.length === 1 && parentKeys[0] === field) {
+            updatedRecipe.source.config[parentField] = undefined;
+        }
+    }
+    return updatedRecipe;
+}
+
+export function setFieldValueOnRecipe(
+    recipe: any,
+    value: string | boolean | null | undefined,
+    field: string,
+    parentField?: string,
```

Comment:
is this assuming that there will be at most 2 layers in the yaml? im concerned that might not always be the case

Reply:
e.g. `profiling.allow_deny_patterns.ignoreCase`

Reply:
or `stateful_ingestion.state_provider.type`

Reply:
could we rename `parentField` to `parentPath`?

Reply:
yeah this was something I thought about but didn't build out right away. At first I was trying to keep it simpler to avoid potential bugs from complexity to ensure things work for the only fields we support at the moment (then thinking make more complex / generalized when needed). However I think just making it more generalized for arbitrary number of layers makes sense to just knock out now and be done with.